### ISSUE TITLE
pcap_dispatch() only returns when there is a lot of packets

### ISSUE THREAD

**Issue by venetwork** [the-tcpdump-group/libpcap#905]
Hello.

"My" code (the most basic example possible):
~~~
#include <pcap.h>
#include <stdio.h>

void monitor(u_char *user, const struct pcap_pkthdr *pkthdr, const u_char *bytes) { return; }

int main(void) {

	pcap_if_t *devs;
	pcap_t *dev;
	char ebuf[PCAP_ERRBUF_SIZE];
	int status;

	pcap_findalldevs(&devs, ebuf);

	dev = pcap_create(devs->name, ebuf);

	if((status = pcap_activate(dev)) < 0) {
		printf("%s\n", pcap_statustostr(status));
		pcap_freealldevs(devs);
		pcap_close(dev);

		return -1;
	}

	for(;;)
		printf("%d\n", pcap_dispatch(dev, 10, monitor, NULL));

	return 0;
}
~~~

doesn't return with a number of captured frames unless I generate a massive amount of packets (downloading an .iso file, etc.)

What's strange is that I clearly remember it working as I'd expect (a single ICMP packet would be captured). I tried changing the maximum amount of frames to catch (1, 20 or 0, etc.) but, again, it only works when the network interface is flooded with traffic.

(edit) Mystery solved. The above code (and my project's code) works fine with libpcap 1.9.0. It would still be nice to have it working with libpcap 1.9.1 and future releases.

Cheers.

**Comment by Oppen** [comment]
What's the platform? Linux?

**Comment by venetwork** [comment]
64-bit Ubuntu 18.04 (all kernels supported by this distribution were tested but it all boiled down to libpcap 1.9.1 not working as expected).

**Comment by guyharris** [comment]
To quote the pcap(3PCAP) man page:

       To  open  a handle for a live capture, given the name of the network or
       other interface on which the capture should  be  done,  call  pcap_cre-
       ate(),  set the appropriate options on the handle, and then activate it
       with pcap_activate().

The code does:

        dev = pcap_create(devs->name, ebuf);

	if((status = pcap_activate(dev)) < 0) {

with no options being set.

This means that it doesn't set a timeout.  This means that the *default* timeout is specified; the default value is 0. Again, quoting pcap(3PCAP):

       packet buffer timeout
              If,  when  capturing,  packets  are  delivered  as  soon as they
              arrive, the application capturing the packets will be  woken  up
              for  each  packet  as  it arrives, and might have to make one or
              more calls to the operating system to fetch each packet.

              If, instead, packets are not delivered as soon as  they  arrive,
              but  are  delivered after a short delay (called a "packet buffer
              timeout"), more than one packet can be  accumulated  before  the
              packets are delivered, so that a single wakeup would be done for
              multiple packets, and each set of calls made  to  the  operating
              system  would  supply  multiple  packets,  rather  than a single
              packet.  This reduces the per-packet CPU overhead if packets are
              arriving  at  a  high rate, increasing the number of packets per
              second that can be captured.

              The packet buffer timeout is required  so  that  an  application
              won't  wait for the operating system's capture buffer to fill up
              before packets are delivered; if packets  are  arriving  slowly,
              that wait could take an arbitrarily long period of time.

              Not  all platforms support a packet buffer timeout; on platforms
              that don't, the packet buffer timeout is ignored.  A zero  value
              for the timeout, on platforms that support a packet buffer time-
              out, will cause a read to wait forever to allow  enough  packets
              to  arrive,  with  no timeout.  A negative value is invalid; the
              result of setting the timeout to  a  negative  value  is  unpre-
              dictable.

              NOTE:  the  packet  buffer timeout cannot be used to cause calls
              that read packets to return within a  limited  period  of  time,
              because, on some platforms, the packet buffer timeout isn't sup-
              ported, and, on other platforms, the timer doesn't  start  until
              at  least one packet arrives.  This means that the packet buffer
              timeout should NOT be  used,  for  example,  in  an  interactive
              application  to  allow  the  packet capture loop to ``poll'' for
              user input periodically, as there's no  guarantee  that  a  call
              reading packets will return after the timeout expires even if no
              packets have arrived.

              The packet buffer timeout is set with pcap_set_timeout().

As it says, "A zero value for the timeout, on platforms that support a packet buffer timeout, will cause a read to wait forever to allow enough packets to arrive, with no timeout."  So it will wait until a bufferful of packets arrives.

> The above code (and my project's code) works fine with libpcap 1.9.0. It would still be nice to have it working with libpcap 1.9.1 and future releases.

As I remember, some people *complained* that it *didn't* wait forever, so that was fixed in 1.9.1.

If you want it to work with 1.9.1 - or on platforms *other* than Linux - set a timeout.  You probably want something no larger than 1 second (1000 milliseconds), possibly .1 (100 milliseconds) or .01 (10 milliseconds).

**Comment by venetwork** [comment]
Thanks for the clarification.
